modules/flowable-engine/src/main/java/org/flowable/engine/impl/interceptor/CommandInvoker.java

```diff
@@ -64,11 +72,36 @@ public void run() {
             executeOperations(commandContext);
     
             // At the end, call the execution tree change listeners.
-            // TODO: optimization: only do this when the tree has actually changed (ie check dbSqlSession).
-            if (!commandContext.isReused() && CommandContextUtil.hasInvolvedExecutions(commandContext)) {
-                agenda.planExecuteInactiveBehaviorsOperation();
+            Map<String, ExecutionEntity> involvedExecutions = CommandContextUtil.getInvolvedExecutions(commandContext);
+            if (!commandContext.isReused() && involvedExecutions != null && !involvedExecutions.isEmpty()) {
+                agenda.planExecuteInactiveBehaviorsOperation(involvedExecutions.values());
+                CommandContextUtil.clearInvolvedExecutions(commandContext);
                 executeOperations(commandContext);
             }
+            
+            VariableListenerSession variableListenerSession = commandContext.getSession(VariableListenerSession.class);
+            Map<String, List<VariableListenerSessionData>> variableSessionData = variableListenerSession.getVariableData();
+            
+            List<String> processInstanceIds = new ArrayList<>();
+            if (variableSessionData != null) {
+                for (String variableName : variableSessionData.keySet()) {
+                    List<VariableListenerSessionData> variableListenerDataList = variableSessionData.get(variableName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.